src/libPMacc/include/random/RNGProvider.tpp

```diff
@@ -0,0 +1,162 @@
+        {
+            const uint32_t linearTid = blockIdx.x * blockDim.x + threadIdx.x;
+            if(linearTid >= size.productOfComponents())
+                return;
+
+            const T_Space cellIdx = DataSpaceOperations<T_Space::dim>::map(size, linearTid);
+            T_RNGMethod().init(rngBox(cellIdx), seed, linearTid);
+        }
+
+        template<class T_Functor, class T_Space>
+        __global__ void
+        testRNGProvider(T_Functor random, const T_Space size)
+        {
+            const uint32_t linearTid = blockIdx.x * blockDim.x + threadIdx.x;
+            if(linearTid >= size.productOfComponents())
+                return;
+
+            const T_Space cellIdx = DataSpaceOperations<T_Space::dim>::map(size, linearTid);
+
+            random.init(cellIdx);
+            printf("%u: %g\n", linearTid, random());
+        }
+
+    }  // namespace kernel
+
+    template<uint32_t T_dim, class T_RNGMethod>
+    RNGProvider<T_dim, T_RNGMethod>::RNGProvider(const Space& size, const std::string& uniqueId):
+    		m_size(size), m_uniqueId(uniqueId.empty() ? getName() : uniqueId),
+    		buffer(new Buffer(size))
+    {
+        if(m_size.productOfComponents() == 0)
+            throw std::invalid_argument("Cannot create RNGProvider with zero size");
+    }
+
+    template<uint32_t T_dim, class T_RNGMethod>
+    void RNGProvider<T_dim, T_RNGMethod>::init(uint32_t seed)
+    {
+
+        const uint32_t blockSize = 256;
+        const uint32_t gridSize = (m_size.productOfComponents() + blockSize - 1) / blockSize; // Round up
+
+#if XRT_USE_MAPPED_RNG_BUFFER
+        auto bufferBox = buffer->getDataBox();
+#else
+        auto bufferBox = buffer->getDeviceBuffer().getDataBox();
+#endif
+        __cudaKernel(kernel::initRNGProvider<RNGMethod>)
+        (gridSize, blockSize)
+        ( bufferBox,
+          seed,
+          m_size
+          );
+
+        Environment<dim>::get().DataConnector().registerData(*this);
+
+#ifdef XRT_DEBUG_RNG
+        PMACC_AUTO(random, createRandom<distributions::Uniform_float<> >());
+        __cudaKernel(kernel::testRNGProvider)
+        (gridSize, blockSize)
+        (random, m_size);
+#endif
+    }
+
+    template<uint32_t T_dim, class T_RNGMethod>
+    typename RNGProvider<T_dim, T_RNGMethod>::Handle
+    RNGProvider<T_dim, T_RNGMethod>::createHandle(const std::string& id)
+    {
+        RNGProvider& provider = Environment<>::get().DataConnector().getData<RNGProvider>(id, true);
+        Handle result(provider.getDeviceDataBox());
+        Environment<>::get().DataConnector().releaseData(id);
+        return result;
+    }
+
+    template<uint32_t T_dim, class T_RNGMethod>
+    template<class T_Distribution>
+    typename RNGProvider<T_dim, T_RNGMethod>::GetRandomType<T_Distribution>::type
+    RNGProvider<T_dim, T_RNGMethod>::createRandom(const std::string& id)
+    {
+        RNGProvider& provider = Environment<>::get().DataConnector().getData<RNGProvider>(id, true);
```

Comment:
The types are different. Random is specialized on the reference. When you construct it from a Handle it contains a pointer, if you construct it from the Provider it inherits from the handle. But Probably the proposed syntax still works (constructing from base class)